File: 49-group-anagrams/group-anagrams.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        List<List<String>> ans = new ArrayList<>();
        Map<String,List<String>> map = new HashMap<>();

        for(String str : strs){
            char[] chr = str.toCharArray();
            Arrays.sort(chr);
            String s = new String(chr);
            if(!map.containsKey(s))
                map.put(s, new ArrayList<>());
            List<String> lst = map.get(s);
            lst.add(str);
            map.put(s,lst);
        }
        for(Map.Entry<String,List<String>> ele : map.entrySet()){
            ans.add(ele.getValue());
        }
        return ans;
    }
}

File: 49-group-anagrams/group-anagrams.kt
class Solution {
    fun groupAnagrams(arr: Array<String>): List<List<String>> {
        val map : HashMap<String,MutableList<String>> = hashMapOf()
        for(ele in arr){
            val key = getString(ele.toCharArray())
            if( !map.containsKey(key) ){
                map.put(key,mutableListOf())
            }
            val list = map[key]!!
            list.add(ele)
            map[key] = list
        }
        return map.values.toList()
    }
    fun getString(char : CharArray) : String{
        val sb = StringBuilder()
        val keyArr = IntArray(26)
        for(ele in char){
            
            keyArr[ele - 'a'] = keyArr[ele - 'a'] + 1
            
        } 
        for(ele in keyArr){
            sb.append(ele).append('#')
        }
        return sb.toString()
    }
}